### PR TITLE
Enhanced Document Verification

### PR DESCRIPTION
**Pull request by jumaallan**
Story: https://app.shortcut.com/smileid/story/10830/add-enhanced-doc-v-support-to-mobile-sdks-v10-android

## Summary

Adds enhanced document verification feature

## Known Issues

Any shortcomings in your work. This may include corner cases not correctly handled or issues related
to but not within the scope of your PR. Design compromises should be discussed here if they were not
already discussed above.

## Test Instructions

Concise test instructions on how to verify that your feature works as intended. This should include
changes to the development environment (if applicable) and all commands needed to run your work.

## Screenshot

If applicable (e.g. UI changes), add screenshots to help explain your work.

### CODE REVIEW

File: lib/src/main/java/com/smileidentity/models/JobStatus.kt [hunk 1]
```diff
@@ -78,7 +90,10 @@ interface JobResult : Parcelable {
     @Parcelize
     value class Freeform(
         val result: String,
-    ) : SmartSelfieJobResult, DocumentVerificationJobResult, BiometricKycJobResult
+    ) : SmartSelfieJobResult,
+        DocumentVerificationJobResult,
+        BiometricKycJobResult,
+        EnhancedDocVJobResult
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/models/JobStatus.kt:96]: EnhancedDocV -> EnhancedDocumentVerification

File: lib/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -148,6 +148,9 @@
     <string name="si_bvn_verification_didnt_receive_otp_at">Didn\'t receive the OTP at %1$s?</string>
     <string name="si_bvn_verification_otp_acronym">OTP</string>
 
+    <!--  Enhanced DocV Screen  -->
+    <string name="si_enhanced_docv_product_name">Enhanced DocV</string>
```
Comment (vanshg) [inline lib/src/main/res/values/strings.xml:152]:
```suggestion
    <string name="si_enhanced_docv_product_name">Enhanced Document Verification</string>
```

File: lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt [hunk 3]
```diff
@@ -273,3 +275,63 @@ fun SmileID.BvnConsentScreen(
...
+    jobId: String = rememberSaveable { randomJobId() },
+    showAttribution: Boolean = true,
+    allowGalleryUpload: Boolean = false,
+    showInstructions: Boolean = true,
+    colorScheme: ColorScheme = SmileID.colorScheme,
+    typography: Typography = SmileID.typography,
+    onResult: SmileIDCallback<EnhancedDocumentVerificationResult> = {},
+) {
+    MaterialTheme(colorScheme = colorScheme, typography = typography) {
+        OrchestratedEnhancedDocumentVerificationScreen(
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt:324]: Is it possible to re-use the existing OrchestratedDocumentVerificationScreen here? i.e. by using EnhancedDocumentVerificationResult as a type parameter maybe?

File: lib/src/main/java/com/smileidentity/models/JobStatus.kt (hunk 1 above)
Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/models/JobStatus.kt:96]: I was renaming the references - Missed this one

File: lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt (hunk 3 above)
Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt:324]: I had that idea initially

Then I thought have different  OrchestratedScreen makes sense, then move the capture components to independent composables

So we re-use those composables in different screens, and have different viewmodels

Comment (vanshg) [reply lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt:324]: My concern with this is that it is a good chunk of repeated code -- from the Fragment to the Orchestration layer to the tests etc

File: lib/src/main/java/com/smileidentity/viewmodel/OrchestratedEnhancedDocVViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import kotlinx.coroutines.launch
+import timber.log.Timber
+import java.io.File
+
+internal data class OrchestratedEnhancedDocVUiState(
+    val currentStep: DocumentCaptureFlow = DocumentCaptureFlow.SelfieCapture,
+    @StringRes val errorMessage: Int? = null,
+)
+
+internal class OrchestratedEnhancedDocVViewModel(
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/viewmodel/OrchestratedEnhancedDocVViewModel.kt:37]: Can you change all the instances of `*DocV*`  to `*DocumentVerification*`? i.e. viewmodel, uistate, etc

File: lib/src/main/java/com/smileidentity/models/EnhancedDocumentVerificationResponse.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+@Parcelize
+@JsonClass(generateAdapter = true)
+data class EnhancedDocumentVerificationResponse(
+    @Json(name = "Actions")
+    val actions: Actions,
+    @Json(name = "Country")
+    val country: String,
+    @Json(name = "DOB")
+    val dOB: String,
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/models/EnhancedDocumentVerificationResponse.kt:16]:
```suggestion
    val dob: String,
```

File: lib/src/main/java/com/smileidentity/models/EnhancedDocumentVerificationResponse.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    @Json(name = "ExpirationDate")
+    val expirationDate: String,
+    @Json(name = "FullName")
+    val fullName: String,
+    @Json(name = "Gender")
+    val gender: String,
+    @Json(name = "IDNumber")
+    val iDNumber: String,
+    @Json(name = "IDType")
+    val iDType: String,
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/models/EnhancedDocumentVerificationResponse.kt:28]:
```suggestion
    val idType: String,
```

File: lib/src/main/java/com/smileidentity/results/SmileIDResult.kt [hunk 7]
```diff
@@ -101,3 +102,9 @@ data class BiometricKycResult(
     val livenessFiles: List<File>,
     val jobStatusResponse: BiometricKycJobStatusResponse,
 ) : Parcelable
+
+@Parcelize
+@JsonClass(generateAdapter = true)
+data class EnhancedDocumentVerificationResult(
+    val jobStatusResponse: EnhancedDocumentVerificationJobStatusResponse,
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/results/SmileIDResult.kt:109]: We should also return selfieFile, documentFrontFile, documentBackFile here

File: sample/src/main/java/com/smileidentity/sample/compose/MainScreen.kt [hunk 8]
```diff
@@ -288,6 +290,35 @@ fun MainScreen(
                         },
                     )
                 }
+                composable(ProductScreen.EnhancedDocumentVerification.route) {
```
Comment (vanshg) [inline sample/src/main/java/com/smileidentity/sample/compose/MainScreen.kt:293]: nit, can we place this code block right under the Document Verification block?

File: sample/src/main/java/com/smileidentity/sample/compose/components/IdSelectorScreen.kt [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+import com.smileidentity.sample.R
+import com.smileidentity.sample.viewmodel.IdTypeSelectorAndFieldInputViewModel
+
+@Composable
+internal fun IdSelectorScreen(
+    modifier: Modifier = Modifier,
+    viewModel: IdTypeSelectorAndFieldInputViewModel = viewModel(),
+    onNext: () -> Unit,
+) {
+    val uiState = viewModel.uiState.collectAsStateWithLifecycle().value
```
Comment (vanshg) [inline sample/src/main/java/com/smileidentity/sample/compose/components/IdSelectorScreen.kt:33]: nit
```suggestion
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
```

File: lib/src/main/java/com/smileidentity/results/SmileIDResult.kt (hunk 7 above)
Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/results/SmileIDResult.kt]: Fixed

**Pull request merged**